Modernize OrientationNotifier

youennf opened the pull request:
#### 81cc3782964dc578acc257645c50d2f6f2c2391e
<pre>
Modernize OrientationNotifier
<a href="https://bugs.webkit.org/show_bug.cgi?id=297656">https://bugs.webkit.org/show_bug.cgi?id=297656</a>
<a href="https://rdar.apple.com/158765573">rdar://158765573</a>

Reviewed by Jean-Yves Avenard.

We remove the need for OrientationNotifier to be a CheckedPtr.
We make OrientationNotifier use a weak hash set to handle observers instead of a Vector as it is simpler and safer to use.
To do so, we make OrientationNotifier::Observer a AbstractThreadSafeRefCountedAndCanMakeWeakPtr and we make RealtimeVideoCaptureSource a OrientationNotifier::Observer instead of its children classes.
A follow-up patch might relocate orientation handling in RealtimeVideoCaptureSource as well as a furuther improvement.

Canonical link: <a href="https://commits.webkit.org/299001@main">https://commits.webkit.org/299001@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fcd03bf4cb0361ffa5efc5bbbb0cd269045b3cdb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117453 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37123 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27743 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123553 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69444 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/dcd7fbe3-7216-4c7e-b632-86ca95536f3c) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119331 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37819 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45710 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89137 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8f35bbf4-4dec-45b7-950a-d2b35d56ceb3) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120405 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30111 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105317 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69646 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e1ca2dfd-aec8-43e9-b00d-263b44458bee) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29167 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23431 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67227 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99516 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23613 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126673 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44350 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33341 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97801 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44708 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101551 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97595 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24841 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42966 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20904 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/40701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44223 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49882 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43680 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47024 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45375 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e8dfce9894fdcece80f2c95026e56b9f178569cc. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49656

jyavenard commented on Source/WebCore/platform/OrientationNotifier.h:
```diff
@@ -26,94 +26,65 @@
+        : m_orientation(orientation)
+    {
+    }
+    ~OrientationNotifier() = default;
```
probably not needed. 
or make it virtual if not above.

jyavenard commented on Source/WebCore/platform/OrientationNotifier.h:
```diff
@@ -26,94 +26,65 @@
 enum class VideoFrameRotation : uint16_t;
 
-class OrientationNotifier final : public CanMakeCheckedPtr<OrientationNotifier> {
+class OrientationNotifier {
```
probably still need to be final, or make the destructor virtual

jyavenard reviewed (approved):
Aren't they all main thread objects only? could have done with just RefCounted/CanMakeWeakRef

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fcd03bf4cb0361ffa5efc5bbbb0cd269045b3cdb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49656

youennf committed:
Modernize OrientationNotifier
https://bugs.webkit.org/show_bug.cgi?id=297656
rdar://158765573

Reviewed by Jean-Yves Avenard.

We remove the need for OrientationNotifier to be a CheckedPtr.
We make OrientationNotifier use a weak hash set to handle observers instead of a Vector as it is simpler and safer to use.
To do so, we make OrientationNotifier::Observer a AbstractThreadSafeRefCountedAndCanMakeWeakPtr and we make RealtimeVideoCaptureSource a OrientationNotifier::Observer instead of its children classes.
A follow-up patch might relocate orientation handling in RealtimeVideoCaptureSource as well as a furuther improvement.

Canonical link: https://commits.webkit.org/299001@main
